File: python/realtimex_sdk/mcp.py
"""
MCP Module - Interact with MCP servers via RealtimeX SDK
"""

from typing import Any, Dict, List, Optional
from .api import ApiModule


class MCPModule(ApiModule):
    """Interact with MCP (Model Context Protocol) servers."""

    def __init__(self, realtimex_url: str, app_id: str, app_name: str = None, api_key: str = None):
        super().__init__(realtimex_url, app_id, app_name, api_key)

    async def get_servers(self, provider: str = "all") -> List[Dict[str, Any]]:
        """
        List configured MCP servers.

        Args:
            provider: Filter by provider - 'local', 'remote', or 'all' (default: 'all')

        Returns:
            List of MCP server dicts with keys: name, display_name, description,
            server_type, enabled, provider, tags
        """
        params = f"?provider={provider}" if provider != "all" else ""
        data = await self._api_call("GET", f"/sdk/mcp/servers{params}")
        return data.get("servers", [])

    async def get_tools(self, server_name: str, provider: str = "local") -> List[Dict[str, Any]]:
        """
        List available tools for a specific MCP server.

        Args:
            server_name: The server name (slug)
            provider: Server provider - 'local' or 'remote' (default: 'local')

        Returns:
            List of tool dicts with keys: name, description, input_schema
        """
        from urllib.parse import quote
        data = await self._api_call(
            "GET",
            f"/sdk/mcp/servers/{quote(server_name, safe='')}/tools?provider={provider}"
        )
        return data.get("tools", [])

    async def execute_tool(
        self,
        server_name: str,
        tool_name: str,
        arguments: Optional[Dict[str, Any]] = None,
        provider: str = "local"
    ) -> Any:
        """
        Execute a tool on an MCP server.

        Args:
            server_name: The server name (slug)
            tool_name: The tool name to execute
            arguments: Arguments to pass to the tool (matches tool's input_schema)
            provider: Server provider - 'local' or 'remote' (default: 'local')

        Returns:
            Tool execution result data
        """
        from urllib.parse import quote
        data = await self._api_call(
            "POST",
            f"/sdk/mcp/servers/{quote(server_name, safe='')}/tools/{quote(tool_name, safe='')}/execute?provider={provider}",
            json={"arguments": arguments or {}}
        )
        return data.get("result")

File: typescript/src/modules/mcp.ts
/**
 * MCP Module - Interact with MCP servers via RealtimeX SDK
 */

import { ApiModule } from './api';

// ── Types ────────────────────────────────────────

export interface MCPServer {
    /** Unique server name (slug) */
    name: string;
    /** User-friendly display name */
    display_name: string;
    /** Server description */
    description: string | null;
    /** Server type: 'stdio', 'http', 'sse', or 'remote' */
    server_type: string;
    /** Whether the server is enabled */
    enabled: boolean;
    /** Provider: 'local' or 'remote' */
    provider: 'local' | 'remote';
    /** Tags / categories */
    tags: string[];
}

export interface MCPTool {
    /** Tool name */
    name: string;
    /** Tool description */
    description: string | null;
    /** JSON Schema describing the tool's input parameters */
    input_schema: Record<string, any>;
}

export interface MCPToolResult {
    /** Whether the execution was successful */
    success: boolean;
    /** Server that executed the tool */
    server: string;
    /** Tool that was executed */
    tool: string;
    /** Provider used */
    provider: string;
    /** Execution result data */
    result: any;
    /** Error message if failed */
    error?: string;
}

// ── Module ───────────────────────────────────────

export class MCPModule extends ApiModule {

    constructor(realtimexUrl: string, appId: string, appName?: string, apiKey?: string) {
        super(realtimexUrl, appId, appName, apiKey);
    }

    /**
     * List configured MCP servers.
     * @param provider - Filter by provider: 'local', 'remote', or 'all' (default: 'all')
     * @returns Array of MCP server objects
     */
    async getServers(provider: 'local' | 'remote' | 'all' = 'all'): Promise<MCPServer[]> {
        const params = provider !== 'all' ? `?provider=${provider}` : '';
        const data = await this.apiCall<{ servers: MCPServer[] }>('GET', `/sdk/mcp/servers${params}`);
        return data.servers;
    }

    /**
     * List available tools for a specific MCP server.
     * @param serverName - The server name (slug)
     * @param provider - Provider: 'local' or 'remote' (default: 'local')
     * @returns Array of tool objects with name, description, and input schema
     */
    async getTools(serverName: string, provider: 'local' | 'remote' = 'local'): Promise<MCPTool[]> {
        const data = await this.apiCall<{ tools: MCPTool[] }>(
            'GET',
            `/sdk/mcp/servers/${encodeURIComponent(serverName)}/tools?provider=${provider}`
        );
        return data.tools;
    }

    /**
     * Execute a tool on an MCP server.
     * @param serverName - The server name (slug)
     * @param toolName - The tool name to execute
     * @param args - Arguments to pass to the tool (matches tool's input_schema)
     * @param provider - Provider: 'local' or 'remote' (default: 'local')
     * @returns Tool execution result
     */
    async executeTool(
        serverName: string,
        toolName: string,
        args: Record<string, any> = {},
        provider: 'local' | 'remote' = 'local'
    ): Promise<any> {
        const data = await this.apiCall<MCPToolResult>(
            'POST',
            `/sdk/mcp/servers/${encodeURIComponent(serverName)}/tools/${encodeURIComponent(toolName)}/execute?provider=${provider}`,
            {
                body: JSON.stringify({ arguments: args }),
            }
        );
        return data.result;
    }
}
